torch/csrc/jit/serialization/source_range_serialization.cpp

```diff
@@ -65,10 +91,36 @@ c10::IValue SourceRangeSerializer::serialize_source(
   }
   c10::intrusive_ptr<c10::ivalue::Tuple> serialized;
   if (s == nullptr) {
-    serialized = c10::ivalue::Tuple::create({"", "", 0});
+    serialized = c10::ivalue::Tuple::create({-1, -1, 0});
   } else {
+    auto text_iter = text_to_pos_.find(s->text());
+    int64_t text_pos = -1;
+    if (text_iter == text_to_pos_.end()) {
+      text_pos = static_cast<int64_t>(texts_.size());
+      IValue text = s->text();
+      text_to_pos_[s->text()] = text_pos;
+      texts_.push_back(text);
+    } else {
+      text_pos = text_iter->second;
+    }
+
+    int64_t fname_pos = -1;
+    if (s->filename().has_value()) {
+      auto fname_iter = text_to_pos_.find(*s->filename());
+      if (fname_iter == text_to_pos_.end()) {
+        fname_pos = static_cast<int64_t>(texts_.size());
+        IValue fname = *s->filename();
+        texts_.push_back(fname);
+        text_to_pos_[texts_.back().toStringView()] = fname_pos;
+      } else {
+        fname_pos = fname_iter->second;
+      }
+    }
+
+    TORCH_INTERNAL_ASSERT(texts_[text_pos] == s->text(), "Text need be equal");
+
     serialized = c10::ivalue::Tuple::create(
-        {s->text(), s->filename(), (int64_t)s->starting_line_no()});
+        {text_pos, fname_pos, (int64_t)s->starting_line_no()});
```

Comment:
Looks like this will dedup source code, could we run a one-time manual test on that ads model to see how much saving we can get?

How about the stack trace part in debug_pkl? Could we compress them too? It looks like a lot of the strings have redundancy.

Reply:
> Looks like this will dedup source code, could we run a one-time manual test on that ads model to see how much saving we can get?
I'll get help from the ads folks to see how to do this.

> How about the stack trace part in debug_pkl? Could we compress them too? It looks like a lot of the strings have redundancy.
yes. SourceView object has string field "text", string field "filename". Sometimes the "text" field is stuffed with source code and sometimes stuffed with stack trace. I initially thought it's always source code until I unpicked and printed out what is in there and saw stack trace.